### ISSUE TITLE
Graceful shutdown for @Async not working in netty (but works in tomcat)

### ISSUE THREAD

**Issue by membersound** [spring-projects/spring-boot#30557]
`spring-boot-2.6.6`

I want my `@Async` tasks to delay a shutdown. The following example works perfectly in a `spring-boot-starter-web` environment with  `tomcat`. But fails in `spring-boot-starter-weblux` `netty` server.

`application.properties`:
```
server.shutdown=graceful
server.shutdown.grace-period=30s
spring.task.execution.shutdown.await-termination=true
```

```
@SpringBootApplication
@EnableAsync
public class AppConfiguration {
	public static void main(String[] args) {
		SpringApplication.run(AppConfiguration.class, args);
	}
}

@Service
public class AsyncService {
	private final Logger LOGGER = LoggerFactory.getLogger(this.getClass());

	@Async
	void runAsync() throws InterruptedException {
		LOGGER.info("Starting async job...");
		TimeUnit.MINUTES.sleep(5);
		LOGGER.info("Async job finished.");
	}
}

**Comment by membersound** [comment]
My bad, I thought that `spring.task.execution.shutdown.await-termination-period` would wait infinite until `server.shutdown.grace-period` catches in.
Setting `spring.task.execution.shutdown.await-termination-period=30s` fixes my problem.

Anyways I'm leaving this open, as this might be a good improvement for the docs that `await-termination-period` may not be unsed if `await-termination=true` is set?

**Comment by pruidong** [comment]
With @EnableAsync, it is recommended that you provide your own asynchronous executor, because @EnableAsync uses SimpleAsyncTaskExecutor by default to execute threads, SimpleAsyncTaskExecutor is a single-threaded thread pool that creates a new thread for each task, and SimpleAsyncTaskExecutor does not limit the number of threads by default. Some In this case, it may lead to memory overflow.

To implement your own asynchronous executor, please refer to: EnableAsync.java documentation (https://docs.spring.io/spring-framework/docs/current/javadoc-api/).

**Comment by snicoll** [comment]
@pruidong thank you for trying to help but that recommendation applies to Spring Framework. Spring Boot [auto-configures](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.task-execution-and-scheduling) a suitable `TaskExecutor` if none is provided so what you've described above does not apply.

**Comment by pruidong** [comment]
> 

sorry. let me change the description.

**Comment by pruidong** [comment]
> @pruidong thank you for trying to help but that recommendation applies to Spring Framework. Spring Boot [auto-configures](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.task-execution-and-scheduling) a suitable `TaskExecutor` if none is provided so what you've described above does not apply.

sorry. let me change the description.

**Comment by wilkinsona** [comment]
I think the current naming of the two properties implies a stronger relationship between `awaitTermination` and `awaitTerminationPeriod` than there actually is. I wonder if we should rename the former to align more closely with `setWaitForTasksToCompleteOnShutdown` on `ExecutorConfigurationSupport` to which it is mapped.

**Comment by ThomazPassarelli** [comment]
Hi there,
 I've got a rename suggestion: `spring.task.execution.shutdown.await-termination` -> `spring.task.execution.shutdown.wait-for-completion`.. I find it more intuitive and also fits the naming style better as suggested in the title.

If you think this suggestion fits the criteria, could you assign the issue to me?

**Comment by wilkinsona** [comment]
Thanks for the suggestion.

We generally try to align the names of our configuration properties with the names of the underlying properties that they configure. There are two such configuration properties here:

1. `spring.task.execution.shutdown.await-termination` that is mapped onto `setWaitForTasksToCompleteOnShutdown`
2. `spring.task.execution.shutdown.await-termination-period` that is mapped onto `setAwaitTerminationMillis`

I think `wait-for-completion` fits quite nicely with that.

Something that we haven't yet considered in this issue are the changes, if any, that should be applied to `TaskExecutorBuilder`. It provides an intermediate layer between the configuration properties and the `ThreadPoolTaskExecutor`. `awaitTermination` on the builder should probably be renamed as well. With virtual threads coming in Java 21, I think it's also worth keeping in mind https://github.com/spring-projects/spring-boot/issues/35711.

In short, I'm not sure that we're ready to proceed here. It feels like we need to do some design work to figure out exactly what should change and how it should change. I'll update the issue's labels accordingly. We'll comment on the issue once that design work has been done. If you have any suggestions for the builder and related classes, they would be welcome.

**Comment by wilkinsona** [comment]
As expected, the introduction of virtual thread support has further complicated things.

### await-termination

With virtual threads, `await-termination` doesn't map onto anything. Instead, it determines whether or not `await-termination-period` is applied.

With platform threads, `await-termination` maps onto `ThreadPoolTaskExecutorBuilder.awaitTermination(boolean)`. It is then passed into `ExecutorConfigurationSupport.setWaitForTasksToCompleteOnShutdown(boolean)`

### await-termination-period

With virtual threads, `await-termination-period`  maps onto `SimpleAsyncTaskExecutorBuilder.taskTerminationTimeout(Duration)` which is then passed into `SimpleAsyncTaskExecutor.setTaskTerminationTimeout(long)`.

With platform threads, `await-termination-period` maps onto `ThreadPoolTaskExecutorBuilder.awaitTerminationPeriod(Duration)` which is then passed into `ExecutorConfigurationSupport.setAwaitTerminationMillis(long)`
